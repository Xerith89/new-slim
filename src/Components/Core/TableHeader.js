import React, { Component } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faSort} from '@fortawesome/free-solid-svg-icons'
import {tableOptions} from '../Util/TasksConfig'
import DisplayTasks from './DisplayTasks'
import {sortType} from '../Util/Enums'

import {sortAscending, sortDescending} from '../Util/Sorting'
import { PropTypes } from 'prop-types';

export default class TableHeader extends Component {
    constructor(props) {
        super(props)

        this.state = {
            sortedTasks: this.props.taskList,
            nextSortTaskName: sortType.ASCENDING,
            nextSortNumber: sortType.ASCENDING,
            nextSortType: sortType.ASCENDING,
            nextSortAssignedTo: sortType.ASCENDING,
            nextSortPriority: sortType.ASCENDING,
            nextsortDueDate: sortType.ASCENDING
        }
    }

    componentDidUpdate(prevProps) {
        if(prevProps.taskList !== this.props.taskList) {
          this.setState({sortedTasks: this.props.taskList,});
        }
      }

    doSort = (propertyName) => {
        const sorted = [...this.state.sortedTasks];
        switch (this.state.nextSortTaskName) {
            case sortType.DEFAULT:
                this.setState({
                   sortedTasks: this.props.taskList,
                   nextSortTaskName: sortType.ASCENDING
                });
                break;
            case sortType.ASCENDING:
                sortAscending(sorted, propertyName);
                this.setState({
                    sortedTasks: sorted,
                    nextSortTaskName: sortType.DESECENDING
                });
                break;
            case sortType.DESECENDING:
                sortDescending(sorted, propertyName);
                this.setState({
                    sortedTasks: sorted,
                    nextSortTaskName: sortType.DEFAULT
                });
                break;
            default:
                break;

        }
    }

    //Switch on button ID to call sort on the correct property
    handleClick = (event) => {
        switch(event) {
            case 0:
                this.doSort('taskName');
                break;
            case 1:
                this.doSort('claimSpecNo');
                break;
            case 2:
                this.doSort('type');
                break;
            case 3:
                this.doSort('assigned');
                break;
            case 4:
                this.doSort('priority');
                break;
            case 5:
                this.doSort('dueDate');
                break;
            default:
                break;
        }    
    }

    render() {
      
        return (
            <React.Fragment>
            <thead data-testid="tableHeader">
                <tr>
                    {tableOptions.map((option,i) => {
                        return (<th scope="col" key={i}>{option} <button id={`id-${i}`} onClick={this.handleClick.bind(this, i)} 
                        style={{border: '0', padding: '0', background: 'none'}} data-toggle="tooltip" data-placement="top" title={`Sort By ${option}`}>
                        <FontAwesomeIcon icon={faSort}/></button></th>)
                    })}
                </tr>
            </thead>
            <tbody>
                {this.props.fetching ? null : <DisplayTasks taskList={this.state.sortedTasks}/>} 
            </tbody>   
            </React.Fragment>
        )
    }
}

TableHeader.propTypes = {
    fetchingTasks: PropTypes.bool,
    taskList: PropTypes.array
}
