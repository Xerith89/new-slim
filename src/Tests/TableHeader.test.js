import React from 'react';
import {render, cleanup, fireEvent} from '@testing-library/react';
import TableHeader from './../Components/Core/TableHeader';
import {shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({adapter: new Adapter()});

afterEach(cleanup);

const  taskList = [
    {
        id: 1,
        taskName: 'Review New Claim',
        claimSpecNo: '5500001',
        type: 'Claim', 
        assignedTo: 'Alex Coupe',
        priority: 'Normal',
        dueDate: '01-09-2019'
    },
    {
        id: 2,
        taskName: 'Review New Invoice',
        claimSpecNo: '5500000',
        type: 'Claim', 
        assignedTo: 'Alex Coupe',
        priority: 'Normal',
        dueDate: '01-09-2019'
    },
    {
        id: 3,
        taskName: 'Diary Review',
        claimSpecNo: '5500244',
        type: 'Claim', 
        assignedTo: 'Donald Duck',
        priority: 'Urgent',
        dueDate: '29-09-2019'
    },
    {
        id: 4,
        taskName: 'Loss Adjustor Report Received',
        claimSpecNo: '5500044',
        type: 'Claim', 
        assignedTo: 'Peter Parker',
        priority: 'Urgent',
        dueDate: '05-09-2019'
    },
]


test('TableHeader Renders To The Dom Without Spinner', () => {
    const {getByTestId} = render(<TableHeader taskList={taskList} fetchingTasks={false}/>);
    expect(getByTestId('tableHeader')).toMatchSnapshot();
});

test('Sorting Function Fires On Click', () => {
    const spy = jest.fn();
    let wrapper = shallow(<TableHeader taskList={taskList} fetchingTasks={false}/>);
    wrapper.instance().handleClick = spy;
    wrapper.instance().forceUpdate();
    wrapper.update();
    expect(wrapper.instance().handleClick).not.toHaveBeenCalled();
    wrapper.find('#id-0').simulate('click');
    expect(wrapper.instance().handleClick).toHaveBeenCalled();
});








