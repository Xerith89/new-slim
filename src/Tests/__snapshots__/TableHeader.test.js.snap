// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableHeader Renders To The Dom Without Spinner 1`] = `
<div
  class="m-3"
  data-testid="tableHeader"
>
  <table
    class="table table-hover table-bordered"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          Name
           
          <button
            data-placement="top"
            data-toggle="tooltip"
            id="id-0"
            style="border: 0px; padding: 0px; background: none;"
            title="Sort By Name"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-sort fa-w-10 "
              data-icon="sort"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                fill="currentColor"
              />
            </svg>
          </button>
        </th>
        <th
          scope="col"
        >
          Spec/Claim No
           
          <button
            data-placement="top"
            data-toggle="tooltip"
            id="id-1"
            style="border: 0px; padding: 0px; background: none;"
            title="Sort By Spec/Claim No"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-sort fa-w-10 "
              data-icon="sort"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                fill="currentColor"
              />
            </svg>
          </button>
        </th>
        <th
          scope="col"
        >
          Type
           
          <button
            data-placement="top"
            data-toggle="tooltip"
            id="id-2"
            style="border: 0px; padding: 0px; background: none;"
            title="Sort By Type"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-sort fa-w-10 "
              data-icon="sort"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                fill="currentColor"
              />
            </svg>
          </button>
        </th>
        <th
          scope="col"
        >
          Assigned
           
          <button
            data-placement="top"
            data-toggle="tooltip"
            id="id-3"
            style="border: 0px; padding: 0px; background: none;"
            title="Sort By Assigned"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-sort fa-w-10 "
              data-icon="sort"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                fill="currentColor"
              />
            </svg>
          </button>
        </th>
        <th
          scope="col"
        >
          Priority
           
          <button
            data-placement="top"
            data-toggle="tooltip"
            id="id-4"
            style="border: 0px; padding: 0px; background: none;"
            title="Sort By Priority"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-sort fa-w-10 "
              data-icon="sort"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                fill="currentColor"
              />
            </svg>
          </button>
        </th>
        <th
          scope="col"
        >
          Due Date
           
          <button
            data-placement="top"
            data-toggle="tooltip"
            id="id-5"
            style="border: 0px; padding: 0px; background: none;"
            title="Sort By Due Date"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-sort fa-w-10 "
              data-icon="sort"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                fill="currentColor"
              />
            </svg>
          </button>
        </th>
      </tr>
    </thead>
    <tbody
      data-testid="tasks"
    >
      <tr>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            Review New Claim
          </button>
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            5500001
          </button>
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            Claim
          </button>
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          />
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            Normal
          </button>
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            01-09-2019
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            Review New Invoice
          </button>
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            5500000
          </button>
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            Claim
          </button>
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          />
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            Normal
          </button>
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            01-09-2019
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            Diary Review
          </button>
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            5500244
          </button>
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            Claim
          </button>
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          />
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            Urgent
          </button>
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            29-09-2019
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            Loss Adjustor Report Received
          </button>
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            5500044
          </button>
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            Claim
          </button>
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          />
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            Urgent
          </button>
        </td>
        <td>
          <button
            class="btn-block"
            style="display: block; border: 0px; padding: 0px; background: none;"
          >
            05-09-2019
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
